src/rulesets/LinearAlgebra/factorization.jl

```diff
@@ -1,6 +1,218 @@
 using LinearAlgebra: checksquare
 using LinearAlgebra.BLAS: gemv, gemv!, gemm!, trsm!, axpy!, ger!
 
+#####
+##### `lu`
+#####
+
+# These rules are necessary because the primals call LAPACK functions
+
+# frule for square matrix was introduced in Eq. 3.6 of
+# de Hoog, F.R., Anderssen, R.S. and Lukas, M.A. (2011)
+# Differentiation of matrix functionals using triangular factorization.
+# Mathematics of Computation, 80 (275). p. 1585.
+# doi: http://doi.org/10.1090/S0025-5718-2011-02451-8
+# for derivations for wide and tall matrices, see
+# https://sethaxen.com/blog/2021/02/differentiating-the-lu-decomposition/
+
+function frule(
+    (_, ΔA), ::typeof(lu!), A::StridedMatrix, pivot::Union{Val{false},Val{true}}; kwargs...
+)
+    F = lu!(A, pivot; kwargs...)
+    ∂factors = pivot === Val(true) ? ΔA[F.p, :] : ΔA
+    m, n = size(∂factors)
+    q = min(m, n)
+    if m == n  # square A
+        # minimal allocation computation of
+        # ∂L = L * tril(L \ (P * ΔA) / U, -1)
+        # ∂U = triu(L \ (P * ΔA) / U) * U
+        # ∂factors = ∂L + ∂U
+        L = UnitLowerTriangular(F.factors)
+        U = UpperTriangular(F.factors)
+        rdiv!(∂factors, U)
+        ldiv!(L, ∂factors)
+        ∂L = lmul!(L, tril(∂factors, -1))
+        ∂U = rmul!(triu(∂factors), U)
+        ∂factors .= ∂L .+ ∂U
+    elseif m < n  # wide A, system is [P*A1 P*A2] = [L*U1 L*U2]
+        L = UnitLowerTriangular(F.L)
+        U = F.U
+        @views begin
+            ∂factors1 = ∂factors[:, 1:q]
+            ∂factors2 = ∂factors[:, (q + 1):n]
+            U1 = UpperTriangular(U[:, 1:q])
+            U2 = U[:, (q + 1):n]
+        end
+        ldiv!(L, ∂factors)
+        rdiv!(∂factors1, U1)
+        ∂L = tril(∂factors1, -1)
+        mul!(∂factors2, ∂L, U2, -1, 1)
+        lmul!(L, ∂L)
+        rmul!(triu!(∂factors1), U1)
+        ∂factors1 .+= ∂L
+        ∂U = triu(∂factors)
+    else  # tall A, system is [P1*A; P2*A] = [L1*U; L2*U]
+        L = F.L
+        U = UpperTriangular(F.U)
+        @views begin
+            ∂factors1 = ∂factors[1:q, :]
+            ∂factors2 = ∂factors[(q + 1):m, :]
+            L1 = UnitLowerTriangular(L[1:q, :])
+            L2 = L[(q + 1):m, :]
+        end
+        rdiv!(∂factors, U)
+        ldiv!(L1, ∂factors1)
+        ∂U = triu(∂factors1)
+        mul!(∂factors2, L2, ∂U, -1, 1)
+        rmul!(∂U, U)
+        lmul!(L1, tril!(∂factors1, -1))
+        ∂factors1 .+= ∂U
+        ∂L = tril(∂factors, -1)
+    end
+    ∂F = Composite{typeof(F)}(; L=∂L, U=∂U, factors=∂factors)
+    return F, ∂F
+end
+
+function rrule(
+    ::typeof(lu), A::StridedMatrix, pivot::Union{Val{false},Val{true}}; kwargs...
+)
+    F = lu(A, pivot; kwargs...)
+    function lu_pullback(ΔF::Composite)
+        ∂L = ΔF.L
+        ∂U = ΔF.U
+        if ∂L isa AbstractZero && ∂U isa AbstractZero
```

Comment:
I like it. But I think this would break type-stability. e.g. someone passes in `Diagonal` `∂U` and `∂L` filled with zeros. This would result in the returned cotangent of `A` being `Diagonal`, whereas if any of the diagonal elements were non-zero, a `Matrix` would be returned.